spatialdata/_core/_rasterize.py

```diff
@@ -0,0 +1,383 @@
+        if "z" in spatial_axes:
+            assert d_to_h_bb is not None
+            target_depth = target_height * d_to_h_bb
+    elif target_height is not None:
+        target_width = target_height * w_to_h_bb
+        if "z" in spatial_axes:
+            assert d_to_h_bb is not None
+            target_depth = target_height * d_to_h_bb
+    elif target_depth is not None:
+        assert d_to_h_bb is not None
+        target_height = target_depth / d_to_h_bb
+        target_width = target_height * w_to_h_bb
+    else:
+        raise RuntimeError("Should not reach here")
+    assert target_width is not None
+    assert isinstance(target_width, float)
+    assert target_height is not None
+    assert isinstance(target_height, float)
+    return target_width, target_height, target_depth
+
+
+@singledispatch
+def rasterize(
+    data: Union[SpatialData, SpatialElement],
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+    target_unit_to_pixels: Optional[float] = None,
+    target_width: Optional[float] = None,
+    target_height: Optional[float] = None,
+    target_depth: Optional[float] = None,
+) -> Union[SpatialData, SpatialImage]:
+    raise RuntimeError("Unsupported type: {type(data)}")
+
+
+@rasterize.register(SpatialData)
+def _(
+    sdata: SpatialData,
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+    target_unit_to_pixels: Optional[float] = None,
+    target_width: Optional[float] = None,
+    target_height: Optional[float] = None,
+    target_depth: Optional[float] = None,
+) -> SpatialData:
+    from spatialdata import SpatialData
+
+    new_images = {}
+    for element_type in ["points", "images", "labels", "shapes"]:
+        elements = getattr(sdata, element_type)
+        for name, element in elements:
+            rasterized = rasterize(
+                data=element,
+                axes=axes,
+                min_coordinate=min_coordinate,
+                max_coordinate=max_coordinate,
+                target_coordinate_system=target_coordinate_system,
+                target_unit_to_pixels=target_unit_to_pixels,
+                target_width=target_width,
+                target_height=target_height,
+                target_depth=target_depth,
+            )
+            new_name = f"{name}_rasterized_{element_type}"
+            new_images[new_name] = rasterized
+    return SpatialData(images=new_images, table=sdata.table)
+
+
+@rasterize.register(SpatialImage)
+@rasterize.register(MultiscaleSpatialImage)
+def _(
+    data: SpatialImage,
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+    target_unit_to_pixels: Optional[float] = None,
+    target_width: Optional[float] = None,
+    target_height: Optional[float] = None,
+    target_depth: Optional[float] = None,
+) -> SpatialImage:
```

Comment:
I think we need a docstring. I'm not sure if it's better here or where the single dispatch function is defined above.

Reply:
I have added the docstring. I have placed it where the single dispatch function is defined.